### PR TITLE
require() should emit assertions

### PR DESCRIPTION
**Pull request by steven-johnson**
require() assumed that calls to halide_error() would never return; this is false, and as a result, 'failing' requires didn't actually fail when used in JIT mode without custom error handler. To accomplish this, Call:require was added as an intrinsic, so Codegen could do the right thing.

Note to reviewers: I modeled the handling of Call:require after Call::if_then_else, but I'm not sure about the usage in Bounds.cpp, nor am I sure whether there are other visitors/mutators that need attention.

### CODE REVIEW

File: test/error/require_fail.cpp [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+    const int kPrime1 = 7829;
+    const int kPrime2 = 7919;
+
+    Buffer<int> result;
+    Param<int> p1, p2;
+    Var x;
+    Func f;
+    f(x) = require((p1 + p2) == kPrime1, 
+                   (p1 + p2) * kPrime2,
+                   "The parameters should add to exactly", (kPrime1 * kPrime2), "but were", p1, p2);
```
Comment (abadams) [inline test/error/require_fail.cpp]: Not that it matters much, but I think the error should use kPrime1, or the condition should use the product.

**Comment by abadams** [comment]
LGTM

**Comment by steven-johnson** [comment]
Buildbot failures that I think are spurious; forcing a general rebuild

**Comment by dsharletg** [comment]
The failing buildbot looks like a legit issue with this change (correctness_require): https://buildbot.halide-lang.org/master/#/builders/52/builds/200/steps/10/logs/stdio sounds like maybe the tests just needs to catch the error and return success (and failure in the absence of an error)?

**Comment by steven-johnson** [comment]
> sounds like maybe the tests just needs to catch the error and return success

Huh, no, that's weird: it's a failure that is getting a different (incorrect) result, but only for the gcc5.3 32-bit builds. That's gonna be fun to track down.

**Comment by steven-johnson** [comment]
Heh: this really never should have worked in the original code; we're expecting the 32-bit float value of 61997851.f to be exactly represented, but that's larger than the 23-bit mantissa in float32, so it gets rounded to 61997852.f; it only "worked" before because the validation value was also rounded, and it "breaks" here because this gcc variant is happening to promote the temporary into a double (which can hold that value).

Using float here was just a testing trick anyway for the previous (broken) implementation, to ensure the necessary casting was present; it's no longer necessary, so I've switched the test to just use int, which should clear things up and be simpler anyway.

**Pull request merged**